### PR TITLE
Reduce search-index.js size

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Coming from:

```
1735683 Dec 16 01:35 build/x86_64-apple-darwin/doc/search-index.js
```

to:

```
727755 Dec 16 01:51 build/x86_64-apple-darwin/doc/search-index.js
```

r? @QuietMisdreavus

### CODE REVIEW

**Review by QuietMisdreavus — APPROVED** [review]
At some point we should probably get more eyes and hands on the minifier crate. I just read over the changes and it looks clever. Doesn't seem to introduce any errors either, at least based on personally generating the std docs and looking in the console. And it gives some impressive results! Great work!

**Comment by QuietMisdreavus** [comment]
@bors r+

(i'd mark it rollup but lockfile changes are kinda problematic to roll up...)

**Comment by kennytm** [comment]
@bors p=104

(rollup fairness)

**Comment by GuillaumeGomez** [comment]
I can't reproduce the errors locally.

@QuietMisdreavus: Any clue about what's going on?

**Comment by QuietMisdreavus** [comment]
Wild guess: Is the tester code getting tripped up by the string values turning from literals to variables?

**Comment by GuillaumeGomez** [comment]
It wouldn't work on other platforms either if it were the case. It just feels like on this specific target, docs aren't completely generated, which doesn't make sense since everything else seems to be working. This is quite frustrating to not being able to see artefacts...

**Comment by GuillaumeGomez** [comment]
Let's go for another try just in case... (still not able to reproduce failure locally).

@bors: r=QuietMisdreavus p=1

**Comment by GuillaumeGomez** [comment]
So on apple-i686, it fails. I really need a VM...

**Comment by GuillaumeGomez** [comment]
Also, why do we have segfaults? O.o

**Comment by kennytm** [comment]
@GuillaumeGomez You could edit `.travis.yml` to make a builder run without requiring r+.

**Comment by mati865** [comment]
I don't think segfaults matter, they are comming from test checking kernel guards (which have to crash). They wouldn't be dispalyed if tests passed but they fail like:
```
[02:24:56] FAILED
[02:24:56] ==> Result not found in 'others': '{"path":"std::string","name":"String"}'
[02:24:56] ==> Result not found in 'others': '{"path":"std::ffi","name":"CString"}'
[02:24:56] ==> Result not found in 'others': '{"path":"std::ffi","name":"OsString"}'
[02:24:56] ==> Result not found in 'in_args': '{"path":"std::str","name":"eq"}'
[02:24:56] ==> Result not found in 'returned': '{"path":"std::string::String","name":"add"}'
```

**Comment by GuillaumeGomez** [comment]
Updated. Should be ready now (at least the errors on travis were fixed).

File: src/librustdoc/html/render.rs [hunk 1]
```diff
@@ -1218,14 +1255,50 @@ fn write_minify(dst: PathBuf, contents: &str, enable_minification: bool) -> Resu
...
+    use minifier::js::{Keyword, ReservedChar, Token};
+
     if enable_minification {
         writeln!(dst, "{}",
-                 minifier::js::minify_and_replace_keywords(contents, keywords_to_replace))
+                 minifier::js::simple_minify(contents)
+                              .apply(|f| {
+                                  // We keep backlines.
+                                  minifier::js::clean_tokens_except(f, |c| {
+                                      c.get_char() != Some(ReservedChar::Backline)
```
Comment (QuietMisdreavus) [inline src/librustdoc/html/render.rs]: The naming of the `clean_tokens_except` function confuses me. Is this call removing everything but newlines?

Comment (GuillaumeGomez) [reply src/librustdoc/html/render.rs]: The story behind it is a bit longer: originally I created a function called `clean_tokens` to remove the useless ones (unwanted backlines, extra whitespaces/tabs, etc...). This function continues to do it but you can make it keep a token (the backline in here) if you want.

**Comment by QuietMisdreavus** [comment]
Let's see if we can get this through bors this time...

@bors r+

**Comment by GuillaumeGomez** [comment]
Nooooooo...

**Comment by QuietMisdreavus** [comment]
*lockfiles...*

r=me when travis is green. Should we bump priority to try to beat any other PRs that touch the lockfile, or hope that the age of this one will let it jump the queue naturally?

**Comment by GuillaumeGomez** [comment]
it's alphabetically sorted iirc so I'll just put a high priority (like 10).

**Comment by QuietMisdreavus** [comment]
```
Checking "substring.js" ... FAILED
==> Exact check failed at position 2: expected '{"path":"std","name":"assert"}' but found '{"crate":"core","ty":14,"name":"debug_assert_eq","path":"core","desc":"Asserts that two expressions are equal to each other.","type":null,"lev":0,"displayPath":"<span>core::</span>","fullPath":"<span>core::</span>debug_assert_eq|14","href":"../core/macro.debug_assert_eq.html"}'
```

**Comment by GuillaumeGomez** [comment]
Ah, exact check might be a problem here.

**Comment by GuillaumeGomez** [comment]
Weird, the PR isn't in the build queue anymore... Let's try again?

@bors: r=QuietMisdreavus p=10

**Comment by GuillaumeGomez** [comment]
And after a rebase, let's go again...

@bors: r=QuietMisdreavus p=10

**Pull request merged**